### PR TITLE
[participant-integration-api] - Separate gRPC command services from the domain command services [KVL-1172]

### PR DESCRIPTION
**Pull request by nicu-da**
### Pull Request Checklist

Best reviewed commit by commit
- Make the command tracker generic so that we can differentiate between internal submission and submission using the gRPC layer
- have the ApiCommandService depend on the ApiSubmissionService and not the gRPC interfaces

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/services/commands/tracker/TrackerCommandInput.scala [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+object Trackable {
+  implicit class TrackableOps[T](value: T)(implicit trackable: Trackable[T]) {
+
+    def trackingInput: TrackerCommandInput = trackable.trackingCommandInput(value)
+
+  }
+}
+
```
Comment (SamirTalwar-DA) [inline ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/services/commands/tracker/TrackerCommandInput.scala:42]: Strange spacing. Can we normalize it a little?

```suggestion

trait Trackable[T] {
  def trackingCommandInput(value: T): TrackerCommandInput
}

object Trackable {
  implicit class TrackableOps[T](value: T)(implicit trackable: Trackable[T]) {
    def trackingInput: TrackerCommandInput = trackable.trackingCommandInput(value)
  }
}
```

File: ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/services/tracking/TrackerMapSpec.scala [hunk 2]
```diff
@@ -44,18 +42,18 @@ class TrackerMapSpec extends AsyncWordSpec with Matchers {
...
 
       for {
         completion1 <- tracker.track(
-          CommandSubmission(Commands(commandId = "1", actAs = Seq("Alice")))
+          "1"
         )
         completion2 <- tracker.track(
-          CommandSubmission(Commands(commandId = "2", actAs = Seq("Bob")))
+          "2"
         )
```
Comment (SamirTalwar-DA) [inline ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/services/tracking/TrackerMapSpec.scala:57]:
```suggestion
        completion1 <- tracker.track("1")
        completion2 <- tracker.track("2")
```

File: ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/services/tracking/TrackerMapSpec.scala [hunk 3]
```diff
@@ -69,32 +67,30 @@ class TrackerMapSpec extends AsyncWordSpec with Matchers {
...
 
       for {
         completion1 <- tracker.track(
-          CommandSubmission(Commands(commandId = "X", actAs = Seq("Alice", "Bob")))
+          Seq("Alice", "Bob")
         )
         completion2 <- tracker.track(
-          CommandSubmission(Commands(commandId = "Y", actAs = Seq("Bob", "Alice")))
+          Seq("Bob", "Alice")
         )
```
Comment (SamirTalwar-DA) [inline ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/services/tracking/TrackerMapSpec.scala:87]: Perhaps join these lines too.

File: ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/services/tracking/TrackerMapSpec.scala [hunk 4]
```diff
@@ -115,7 +111,7 @@ class TrackerMapSpec extends AsyncWordSpec with Matchers {
 
       val requests = (0 until requestCount).map { i =>
         val key = (i % expectedTrackerCount).toString
```
Comment (SamirTalwar-DA) [inline ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/services/tracking/TrackerMapSpec.scala:113]: You can inline this.

File: ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/services/tracking/TrackerMapSpec.scala [hunk 5]
```diff
@@ -217,7 +213,7 @@ class TrackerMapSpec extends AsyncWordSpec with Matchers {
...
```
Comment (SamirTalwar-DA) [inline ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/services/tracking/TrackerMapSpec.scala:215]: And here.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiCommandService.scala [hunk 6]
```diff
@@ -80,118 +82,97 @@ private[apiserver] final class ApiCommandService private[services] (
...
-      request: SubmitAndWaitRequest
+      request: SubmitRequest,
+      trackingTimeout: Option[Duration],
   ): Future[SubmitAndWaitForTransactionResponse] =
-    withCommandsLoggingContext(request.getCommands) { case (enrichedLoggingContext, errorLogger) =>
-      submitAndWaitInternal(request)(enrichedLoggingContext, errorLogger).flatMap {
-        case Left(failure) =>
-          Future.failed(CompletionResponse.toException(failure, errorFactories)(errorLogger))
-        case Right(resp) =>
-          val effectiveActAs = CommandsValidator.effectiveSubmitters(request.getCommands).actAs
```
Comment (SamirTalwar-DA) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiCommandService.scala:118]: Do we not need to do this any more?

File: ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/services/ApiCommandServiceSpec.scala [hunk 7]
```diff
@@ -148,20 +98,17 @@ class ApiCommandServiceSpec
             )
           )
         )
-
-        openChannel(
+        val service =
           new ApiCommandService(
```
Comment (SamirTalwar-DA) [inline ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/services/ApiCommandServiceSpec.scala:102]: You can join these two lines.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiCommandService.scala [hunk 8]
```diff
@@ -225,7 +206,8 @@ private[apiserver] object ApiCommandService {
     val ledgerOffsetValidator = new LedgerOffsetValidator(errorFactories)
     val submissionTracker = new TrackerMap.SelfCleaning(
       configuration.trackerRetentionPeriod,
-      Tracking.getTrackerKey,
+      (internalSubmission: InternalCommandSubmission) =>
+        Tracking.getTrackerKey(internalSubmission.request.commands),
```
Comment (SamirTalwar-DA) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiCommandService.scala:210]: Can we add this to the `Trackable` typeclass?

**Review by SamirTalwar-DA — COMMENTED** [review]
I know this is still in progress, but I found it helpful to review it anyway.

**Comment by SamirTalwar-DA** [comment]
@nicu-da, is this still in progress?

**Comment by nicu-da** [comment]
@SamirTalwar-DA 
I'm thinking of dropping it. 
While it should be easy to update to the latest main, as far as I'm aware the participant team is planning on refactoring this area anyway so it might not be that helpful. 
It was a good exercise to get some knowledge in this part of the code at least.

**Comment by SamirTalwar-DA** [comment]
Makes sense. Thanks for the update.

**Pull request closed without merging**